src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -0,0 +1,1417 @@
+
+            if (!(validatedAccessToken is JsonWebToken jwtValidatedAccessToken))
+                throw LogHelper.LogExceptionMessage(new SignedHttpRequestValidationException(LogHelper.FormatInvariant(LogMessages.IDX23031, validatedAccessToken.GetType(), typeof(JsonWebToken), validatedAccessToken)));
+
+            // use the decrypted jwt if the jwtValidatedAccessToken is encrypted.
+            if (jwtValidatedAccessToken.InnerToken != null)
+                jwtValidatedAccessToken = jwtValidatedAccessToken.InnerToken;
+            
+            if (jwtValidatedAccessToken.TryGetPayloadValue(ConfirmationClaimTypes.Cnf, out JObject cnf) && cnf != null)
+                return cnf.ToString();
+            else if (jwtValidatedAccessToken.TryGetPayloadValue(ConfirmationClaimTypes.Cnf, out string cnfString) && !string.IsNullOrEmpty(cnfString))
+                return cnfString;
+            else
+                throw LogHelper.LogExceptionMessage(new SignedHttpRequestInvalidCnfClaimException(LogHelper.FormatInvariant(LogMessages.IDX23003, ConfirmationClaimTypes.Cnf)));
+        }
+
+        /// <summary>
+        /// Resolves a PoP <see cref="SecurityKey"/> from a confirmation ("cnf") claim.
+        /// </summary>
+        /// <param name="confirmationClaim">A confirmation ("cnf") claim as a string.</param>
+        /// <param name="signedHttpRequest">A signed http request as a JWT.</param>
+        /// <param name="validatedAccessToken">An access token ("at") that was already validated during the SignedHttpRequest validation process.</param>
+        /// <param name="signedHttpRequestValidationContext">A structure that wraps parameters needed for SignedHttpRequest validation.</param>
+        /// <param name="cancellationToken">Propagates notification that operations should be canceled.</param>
+        /// <returns>A resolved PoP <see cref="SecurityKey"/>.</returns>
+        /// <remarks>https://tools.ietf.org/html/rfc7800#section-3.1</remarks>
+        protected virtual async Task<SecurityKey> ResolvePopKeyFromCnfClaimAsync(string confirmationClaim, SecurityToken signedHttpRequest, SecurityToken validatedAccessToken, SignedHttpRequestValidationContext signedHttpRequestValidationContext, CancellationToken cancellationToken)
+        {
+            if (string.IsNullOrEmpty(confirmationClaim))
+                throw LogHelper.LogArgumentNullException(nameof(confirmationClaim));
+
+            var cnf = JObject.Parse(confirmationClaim);
+            if (cnf.TryGetValue(ConfirmationClaimTypes.Jwk, StringComparison.Ordinal, out var jwk))
+            {
+                return ResolvePopKeyFromJwk(jwk.ToString(), signedHttpRequest, validatedAccessToken, signedHttpRequestValidationContext);
+            }
+            else if (cnf.TryGetValue(ConfirmationClaimTypes.Jwe, StringComparison.Ordinal, out var jwe))
+            {
+                return await ResolveP
```

Comment:
Can we put this in a utility class?

Reply:
Yes, makes sense. Will do.

Reply:
Addressed with https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/commit/0fac2cc80a8875dea87c068ef99978ca51ad68ba